api/tree.go

```diff
@@ -124,19 +126,39 @@ func (s *Server) CreateTree(w http.ResponseWriter, r *http.Request) {
 		Addr  string   `json:"addr"`
 		Proof []string `json:"proof"`
 	}
-	var proofs = []proofItem{}
+	var (
+		proofs = []proofItem{}
+		eg     errgroup.Group
+		pm     sync.Mutex
+	)
 	for _, l := range leaves {
-		pf := tree.Proof(l)
-		if !merkle.Valid(tree.Root(), pf, l) {
-			s.sendJSONError(r, w, nil, http.StatusBadRequest, "Unable to generate proof for tree")
-			return
-		}
-		proofs = append(proofs, proofItem{
-			Leaf:  hexutil.Encode(l),
-			Addr:  leaf2Addr(l, req.Ltd, req.Packed).Hex(),
-			Proof: encodeProof(pf),
+		l := l //avoid capture
+		eg.Go(func() error {
+			pf := tree.Proof(l)
+			if !merkle.Valid(tree.Root(), pf, l) {
+				return errors.New("invalid proof for tree")
+			}
+			var (
+				addr         = leaf2Addr(l, req.Ltd, req.Packed).Hex()
+				encodedLeaf  = hexutil.Encode(l)
+				encodedProof = encodeProof(pf)
+			) // encode outside of lock
+			pm.Lock()
+			proofs = append(proofs, proofItem{
```

Comment:
Asking mostly for my own understanding: Do we lose anything because the proofs are now being added in a non deterministic order? I imagine this is fine because we are saving the tree in the original order

Reply:
The proof is ordered but I don't think the proofs need to be ordered.